File: apps/static-site/src/content/docs/0_start/schema.ts
// schemas.ts
import { z } from "zod";

import { createPostgresPublicSchemaTransformations } from "@rejot-dev/adapter-postgres";
import { createConsumerSchema } from "@rejot-dev/contract/consumer-schema";
import { createPublicSchema } from "@rejot-dev/contract/public-schema";

// Public schema definition for api_key table
const apiKeyPublicSchema = createPublicSchema("my-public-schema", {
  source: { dataStoreSlug: "my-db" },
  outputSchema: z.object({
    id: z.string(),
    api_key: z.string(),
  }),
  config: {
    publicSchemaType: "postgres",
    transformations: [
      ...createPostgresPublicSchemaTransformations(
        "insertOrUpdate",
        "api_key",
        `SELECT id, key AS "api_key" FROM api_key WHERE id = :id`,
      ),
    ],
  },
  version: {
    major: 1,
    minor: 0,
  },
});

// Consumer schema that writes to target_table
const apiKeyConsumerSchema = createConsumerSchema("my-consumer-schema", {
  source: {
    manifestSlug: "my-sync-project",
    publicSchema: {
      name: "my-public-schema",
      majorVersion: 1,
    },
  },
  config: {
    consumerSchemaType: "postgres",
    destinationDataStoreSlug: "my-db",
    sql: `INSERT INTO target_table (id, api_key) VALUES (:id, :api_key)
          ON CONFLICT (id) DO UPDATE SET api_key = :api_key`,
  },
});

export default {
  apiKeyPublicSchema,
  apiKeyConsumerSchema,
};

File: apps/static-site/src/content/docs/0_start/schema.py
# schema.py
from pydantic import BaseModel
from rejot_contract.public_schema import (
    create_public_schema, PublicSchemaConfig, Source, Version, create_postgres_public_schema_transformation
)
from rejot_contract.consumer_schema import (
    create_consumer_schema, ConsumerSchemaConfig, SourceManifest, PublicSchema
)

class ApiKey(BaseModel):
    id: int
    api_key: str


public = create_public_schema(
    public_schema_name="my-public-schema",
    source=Source(dataStoreSlug="my-db"),
    output_schema=ApiKey,
    version=Version(major=1, minor=0),
    config=PublicSchemaConfig(
        publicSchemaType="postgres",
        transformations=[
            create_postgres_public_schema_transformation(
                operation="insertOrUpdate",
                table="api_key",
                sql="SELECT id, key AS 'api_key' FROM api_key WHERE id = :id",
            ),
        ],
    ),
)

# Define a consumer schema
consumer = create_consumer_schema(
    "my-consumer-schema",
    source=SourceManifest(
        manifestSlug="my-sync-project",
        publicSchema=PublicSchema(
            name="my-public-schema",
            majorVersion=1,
        ),
    ),
    config=ConsumerSchemaConfig(
        consumerSchemaType="postgres",
        destinationDataStoreSlug="my-db",
        sql="INSERT INTO target_table (id, api_key) VALUES (:id, :api_key) ON CONFLICT (id) DO UPDATE SET api_key = :api_key",
    ),
)
